File: ios/Proxy/CMCDProxy.swift
import Foundation
import Network
import ExpoModulesCore

/// A lightweight local HTTP proxy server for injecting CMCD headers into video requests.
/// This proxy intercepts requests, adds dynamic CMCD headers, and forwards them to the CDN.
///
/// Architecture:
/// ```
/// AVPlayer → localhost:PORT/proxy?url=ENCODED_URL → CMCDProxy → CDN
///                                                      ↓
///                                               + CMCD Headers
/// ```
@available(iOS 13.0, *)
final class CMCDProxy: @unchecked Sendable {

  // MARK: - Properties

  private var listener: NWListener?
  private var connections: [NWConnection] = []
  private let queue = DispatchQueue(label: "expo.video.cmcd.proxy", qos: .userInitiated)
  private let connectionsLock = NSLock()

  private(set) var port: UInt16 = 0
  private(set) var isRunning: Bool = false

  /// Closure to get current CMCD headers - called for each request
  var cmcdHeadersProvider: (() -> [String: String])?

  /// Static headers to add to all requests (e.g., authorization)
  var staticHeaders: [String: String] = [:]

  // MARK: - Lifecycle

  /// Starts the proxy server on a random available port
  func start() throws {
    guard !isRunning else { return }

    // Create TCP listener on random port
    let parameters = NWParameters.tcp
    parameters.allowLocalEndpointReuse = true

    listener = try NWListener(using: parameters, on: .any)

    listener?.stateUpdateHandler = { [weak self] state in
      switch state {
      case .ready:
        if let port = self?.listener?.port?.rawValue {
          self?.port = port
          self?.isRunning = true
          log.info("[CMCDProxy] Started on port \(port)")
        }
      case .failed(let error):
        log.error("[CMCDProxy] Failed to start: \(error)")
        self?.isRunning = false
      case .cancelled:
        self?.isRunning = false
        log.info("[CMCDProxy] Stopped")
      default:
        break
      }
    }

    listener?.newConnectionHandler = { [weak self] connection in
      self?.handleConnection(connection)
    }

    listener?.start(queue: queue)
  }

  /// Stops the proxy server
  func stop() {
    listener?.cancel()
    listener = nil

    connectionsLock.lock()
    connections.forEach { $0.cancel() }
    connections.removeAll()
    connectionsLock.unlock()

    isRunning = false
    port = 0
  }

  // MARK: - Connection Handling

  private func handleConnection(_ connection: NWConnection) {
    connectionsLock.lock()
    connections.append(connection)
    connectionsLock.unlock()

    connection.stateUpdateHandler = { [weak self, weak connection] state in
      if case .cancelled = state, let conn = connection {
        self?.removeConnection(conn)
      }
    }

    connection.start(queue: queue)
    receiveRequest(from: connection)
  }

  private func removeConnection(_ connection: NWConnection) {
    connectionsLock.lock()
    connections.removeAll { $0 === connection }
    connectionsLock.unlock()
  }

  private func receiveRequest(from connection: NWConnection) {
    connection.receive(minimumIncompleteLength: 1, maximumLength: 65536) { [weak self] data, _, isComplete, error in
      guard let self = self, let data = data, !data.isEmpty else {
        connection.cancel()
        return
      }

      self.processRequest(data: data, connection: connection)
    }
  }

  // MARK: - Request Processing

  private func processRequest(data: Data, connection: NWConnection) {
    guard let requestString = String(data: data, encoding: .utf8) else {
      sendErrorResponse(to: connection, statusCode: 400, message: "Bad Request")
      return
    }

    // Parse HTTP request
    let lines = requestString.components(separatedBy: "\r\n")
    guard let requestLine = lines.first else {
      sendErrorResponse(to: connection, statusCode: 400, message: "Bad Request")
      return
    }

    let parts = requestLine.components(separatedBy: " ")
    guard parts.count >= 2 else {
      sendErrorResponse(to: connection, statusCode: 400, message: "Bad Request")
      return
    }

    let method = parts[0]
    let path = parts[1]

    // Parse request headers
    var requestHeaders: [String: String] = [:]
    for line in lines.dropFirst() {
      if line.isEmpty { break }
      if let colonIndex = line.firstIndex(of: ":") {
        let key = String(line[..<colonIndex]).trimmingCharacters(in: .whitespaces)
        let value = String(line[line.index(after: colonIndex)...]).trimmingCharacters(in: .whitespaces)
        requestHeaders[key] = value
      }
    }

    // Extract target URL from path
    // Expected format: /proxy?url=ENCODED_URL or /proxy/ENCODED_PATH
    guard let targetUrl = extractTargetUrl(from: path) else {
      sendErrorResponse(to: connection, statusCode: 400, message: "Missing target URL")
      return
    }

    // Proxy the request
    proxyRequest(
      method: method,
      targetUrl: targetUrl,
      originalHeaders: requestHeaders,
      connection: connection
    )
  }

  private func extractTargetUrl(from path: String) -> URL? {
    // Format: /proxy?url=ENCODED_URL
    guard let urlComponents = URLComponents(string: path),
          let queryItems = urlComponents.queryItems,
          let urlParam = queryItems.first(where: { $0.name == "url" })?.value,
          let decodedUrl = urlParam.removingPercentEncoding,
          let url = URL(string: decodedUrl) else {
      return nil
    }
    return url
  }

  private func proxyRequest(
    method: String,
    targetUrl: URL,
    originalHeaders: [String: String],
    connection: NWConnection
  ) {
    var request = URLRequest(url: targetUrl)
    request.httpMethod = method
    request.timeoutInterval = 30

    // Copy relevant headers from original request
    let headersToCopy = ["Range", "Accept", "Accept-Encoding", "Accept-Language"]
    for header in headersToCopy {
      if let value = originalHeaders[header] {
        request.setValue(value, forHTTPHeaderField: header)
      }
    }

    // Add static headers
    for (key, value) in staticHeaders {
      request.setValue(value, forHTTPHeaderField: key)
    }

    // Add dynamic CMCD headers
    if let cmcdHeaders = cmcdHeadersProvider?() {
      for (key, value) in cmcdHeaders {
        request.setValue(value, forHTTPHeaderField: key)
      }
      log.info("[CMCDProxy] Added CMCD headers to request: \(targetUrl.lastPathComponent)")
    }

    // Perform the request
    let task = URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
      if let error = error {
        log.error("[CMCDProxy] Request failed: \(error.localizedDescription)")
        self?.sendErrorResponse(to: connection, statusCode: 502, message: "Bad Gateway")
        return
      }

      guard let httpResponse = response as? HTTPURLResponse,
            let data = data else {
        self?.sendErrorResponse(to: connection, statusCode: 502, message: "Bad Gateway")
        return
      }

      self?.sendProxiedResponse(
        to: connection,
        statusCode: httpResponse.statusCode,
        headers: httpResponse.allHeaderFields as? [String: String] ?? [:],
        body: data,
        originalUrl: targetUrl
      )
    }
    task.resume()
  }

  // MARK: - Response Handling

  private func sendProxiedResponse(
    to connection: NWConnection,
    statusCode: Int,
    headers: [String: String],
    body: Data,
    originalUrl: URL
  ) {
    var responseHeaders = headers

    // Check if this is an HLS manifest that needs URL rewriting
    let contentType = headers["Content-Type"]?.lowercased() ?? ""
    var responseBody = body

    if contentType.contains("mpegurl") || contentType.contains("x-mpegurl") {
      // Rewrite manifest URLs to go through our proxy
      responseBody = rewriteManifest(data: body, baseUrl: originalUrl)
    }

    // Build HTTP response
    var response = "HTTP/1.1 \(statusCode) \(httpStatusMessage(statusCode))\r\n"

    // Add headers
    response += "Content-Length: \(responseBody.count)\r\n"
    response += "Connection: close\r\n"

    // Copy relevant response headers
    let headersToInclude = ["Content-Type", "Cache-Control", "Accept-Ranges", "Content-Range"]
    for header in headersToInclude {
      if let value = responseHeaders[header] {
        response += "\(header): \(value)\r\n"
      }
    }

    // Add CORS headers for flexibility
    response += "Access-Control-Allow-Origin: *\r\n"
    response += "\r\n"

    // Combine headers and body
    var fullResponse = Data(response.utf8)
    fullResponse.append(responseBody)

    connection.send(content: fullResponse, completion: .contentProcessed { _ in
      connection.cancel()
    })
  }

  private func sendErrorResponse(to connection: NWConnection, statusCode: Int, message: String) {
    let body = message.data(using: .utf8) ?? Data()
    let response = """
      HTTP/1.1 \(statusCode) \(message)\r
      Content-Type: text/plain\r
      Content-Length: \(body.count)\r
      Connection: close\r
      \r
      \(message)
      """

    connection.send(content: Data(response.utf8), completion: .contentProcessed { _ in
      connection.cancel()
    })
  }

  private func httpStatusMessage(_ code: Int) -> String {
    switch code {
    case 200: return "OK"
    case 206: return "Partial Content"
    case 301: return "Moved Permanently"
    case 302: return "Found"
    case 304: return "Not Modified"
    case 400: return "Bad Request"
    case 403: return "Forbidden"
    case 404: return "Not Found"
    case 500: return "Internal Server Error"
    case 502: return "Bad Gateway"
    default: return "Unknown"
    }
  }

  // MARK: - Manifest Rewriting

  /// Rewrites URLs in HLS manifest to route through the proxy
  private func rewriteManifest(data: Data, baseUrl: URL) -> Data {
    guard let manifest = String(data: data, encoding: .utf8) else {
      return data
    }

    let proxyBase = "http://127.0.0.1:\(port)/proxy?url="
    var rewrittenLines: [String] = []
    let lines = manifest.components(separatedBy: "\n")

    for line in lines {
      let trimmed = line.trimmingCharacters(in: .whitespaces)

      // Empty lines - preserve
      if trimmed.isEmpty {
        rewrittenLines.append(line)
        continue
      }

      // Handle URI attributes in tags
      if trimmed.hasPrefix("#") && trimmed.contains("URI=\"") {
        let rewritten = rewriteUriAttributes(in: line, baseUrl: baseUrl, proxyBase: proxyBase)
        rewrittenLines.append(rewritten)
        continue
      }

      // Regular comment/tag lines - preserve
      if trimmed.hasPrefix("#") {
        rewrittenLines.append(line)
        continue
      }

      // URL line - rewrite to use proxy
      if let proxyUrl = createProxyUrl(for: trimmed, baseUrl: baseUrl, proxyBase: proxyBase) {
        rewrittenLines.append(proxyUrl)
      } else {
        rewrittenLines.append(line)
      }
    }

    return rewrittenLines.joined(separator: "\n").data(using: .utf8) ?? data
  }

  private func rewriteUriAttributes(in line: String, baseUrl: URL, proxyBase: String) -> String {
    var result = line
    let pattern = "URI=\"([^\"]+)\""

    guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else {
      return line
    }

    let range = NSRange(line.startIndex..., in: line)
    let matches = regex.matches(in: line, options: [], range: range)

    for match in matches.reversed() {
      guard let uriRange = Range(match.range(at: 1), in: line) else { continue }
      let uriValue = String(line[uriRange])

      if let proxyUrl = createProxyUrl(for: uriValue, baseUrl: baseUrl, proxyBase: proxyBase) {
        let fullMatchRange = Range(match.range, in: result)!
        result.replaceSubrange(fullMatchRange, with: "URI=\"\(proxyUrl)\"")
      }
    }

    return result
  }

  private func createProxyUrl(for urlString: String, baseUrl: URL, proxyBase: String) -> String? {
    let trimmed = urlString.trimmingCharacters(in: .whitespaces)

    // Resolve to absolute URL
    let absoluteUrl: URL?
    if trimmed.hasPrefix("http://") || trimmed.hasPrefix("https://") {
      absoluteUrl = URL(string: trimmed)
    } else {
      absoluteUrl = URL(string: trimmed, relativeTo: baseUrl)?.absoluteURL
    }

    guard let url = absoluteUrl,
          let encoded = url.absoluteString.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) else {
      return nil
    }

    return proxyBase + encoded
  }
}

File: android/src/main/java/expo/modules/video/proxy/CMCDProxy.kt
package expo.modules.video.proxy

import android.util.Log
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStream
import java.net.ServerSocket
import java.net.Socket
import java.net.URI
import java.net.URLDecoder
import java.net.URLEncoder
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

/**
 * A lightweight local HTTP proxy server for injecting CMCD headers into video requests.
 * This proxy intercepts requests, adds dynamic CMCD headers, and forwards them to the CDN.
 *
 * Architecture:
 * ```
 * ExoPlayer → localhost:PORT/proxy?url=ENCODED_URL → CMCDProxy → CDN
 *                                                       ↓
 *                                                + CMCD Headers
 * ```
 */
class CMCDProxy {

  companion object {
    private const val TAG = "CMCDProxy"
    private const val BUFFER_SIZE = 8192
  }

  // MARK: - Properties

  private var serverSocket: ServerSocket? = null
  private var executor: ExecutorService? = null
  private val running = AtomicBoolean(false)
  private val httpClient = OkHttpClient.Builder()
    .connectTimeout(30, TimeUnit.SECONDS)
    .readTimeout(30, TimeUnit.SECONDS)
    .writeTimeout(30, TimeUnit.SECONDS)
    .build()

  var port: Int = 0
    private set

  val isRunning: Boolean
    get() = running.get()

  /** Provider for dynamic CMCD headers - called for each request */
  private val cmcdHeadersProvider = AtomicReference<(() -> Map<String, String>)?>(null)

  /** Static headers to add to all requests */
  private val staticHeaders = AtomicReference<Map<String, String>>(emptyMap())

  // MARK: - Lifecycle

  /**
   * Starts the proxy server on an available port.
   * @throws Exception if the server fails to start
   */
  @Throws(Exception::class)
  fun start() {
    if (running.get()) {
      Log.i(TAG, "Proxy already running on port $port")
      return
    }

    // Find available port
    serverSocket = ServerSocket(0).apply {
      reuseAddress = true
    }
    port = serverSocket!!.localPort

    executor = Executors.newCachedThreadPool()
    running.set(true)

    // Start accepting connections
    executor?.submit {
      acceptConnections()
    }

    Log.i(TAG, "Started on port $port")
  }

  /**
   * Stops the proxy server.
   */
  fun stop() {
    running.set(false)

    try {
      serverSocket?.close()
    } catch (e: Exception) {
      Log.e(TAG, "Error closing server socket", e)
    }

    executor?.shutdown()
    try {
      executor?.awaitTermination(5, TimeUnit.SECONDS)
    } catch (e: InterruptedException) {
      executor?.shutdownNow()
    }

    serverSocket = null
    executor = null
    port = 0

    Log.i(TAG, "Stopped")
  }

  /**
   * Sets the CMCD headers provider closure.
   */
  fun setCmcdHeadersProvider(provider: () -> Map<String, String>) {
    cmcdHeadersProvider.set(provider)
  }

  /**
   * Sets static headers to add to all requests.
   */
  fun setStaticHeaders(headers: Map<String, String>) {
    staticHeaders.set(headers)
  }

  // MARK: - Connection Handling

  private fun acceptConnections() {
    while (running.get()) {
      try {
        val socket = serverSocket?.accept() ?: break
        executor?.submit { handleConnection(socket) }
      } catch (e: Exception) {
        if (running.get()) {
          Log.e(TAG, "Error accepting connection", e)
        }
      }
    }
  }

  private fun handleConnection(socket: Socket) {
    try {
      socket.soTimeout = 30000

      val reader = BufferedReader(InputStreamReader(socket.getInputStream()))
      val output = socket.getOutputStream()

      // Read request line
      val requestLine = reader.readLine() ?: return
      val parts = requestLine.split(" ")
      if (parts.size < 2) {
        sendError(output, 400, "Bad Request")
        return
      }

      val method = parts[0]
      val path = parts[1]

      // Read headers
      val requestHeaders = mutableMapOf<String, String>()
      var line = reader.readLine()
      while (!line.isNullOrEmpty()) {
        val colonIndex = line.indexOf(':')
        if (colonIndex > 0) {
          val key = line.substring(0, colonIndex).trim()
          val value = line.substring(colonIndex + 1).trim()
          requestHeaders[key] = value
        }
        line = reader.readLine()
      }

      // Extract target URL
      val targetUrl = extractTargetUrl(path)
      if (targetUrl == null) {
        sendError(output, 400, "Missing target URL")
        return
      }

      // Proxy the request
      proxyRequest(method, targetUrl, requestHeaders, output)

    } catch (e: Exception) {
      Log.e(TAG, "Error handling connection", e)
    } finally {
      try {
        socket.close()
      } catch (e: Exception) {
        // Ignore
      }
    }
  }

  private fun extractTargetUrl(path: String): String? {
    // Format: /proxy?url=ENCODED_URL
    val uri = try {
      URI(path)
    } catch (e: Exception) {
      return null
    }

    val query = uri.query ?: return null
    val params = query.split("&")

    for (param in params) {
      val keyValue = param.split("=", limit = 2)
      if (keyValue.size == 2 && keyValue[0] == "url") {
        return try {
          URLDecoder.decode(keyValue[1], "UTF-8")
        } catch (e: Exception) {
          null
        }
      }
    }

    return null
  }

  // MARK: - Request Proxying

  private fun proxyRequest(
    method: String,
    targetUrl: String,
    originalHeaders: Map<String, String>,
    output: OutputStream
  ) {
    try {
      val requestBuilder = Request.Builder()
        .url(targetUrl)
        .method(method, null)

      // Copy relevant headers
      val headersToCopy = listOf("Range", "Accept", "Accept-Encoding", "Accept-Language")
      for (header in headersToCopy) {
        originalHeaders[header]?.let { requestBuilder.addHeader(header, it) }
      }

      // Add static headers
      for ((key, value) in staticHeaders.get()) {
        requestBuilder.addHeader(key, value)
      }

      // Add dynamic CMCD headers
      cmcdHeadersProvider.get()?.invoke()?.let { cmcdHeaders ->
        for ((key, value) in cmcdHeaders) {
          requestBuilder.addHeader(key, value)
        }
        Log.d(TAG, "Added CMCD headers to request: ${targetUrl.substringAfterLast('/')}")
      }

      val response = httpClient.newCall(requestBuilder.build()).execute()

      // Check if this is an HLS manifest that needs rewriting
      val contentType = response.header("Content-Type")?.lowercase() ?: ""
      var body = response.body?.bytes() ?: ByteArray(0)

      if (contentType.contains("mpegurl") || contentType.contains("x-mpegurl")) {
        body = rewriteManifest(body, targetUrl)
      }

      // Send response
      sendResponse(
        output = output,
        statusCode = response.code,
        statusMessage = response.message,
        headers = response.headers.toMap(),
        body = body
      )

      response.close()

    } catch (e: Exception) {
      Log.e(TAG, "Error proxying request", e)
      sendError(output, 502, "Bad Gateway")
    }
  }

  // MARK: - Response Handling

  private fun sendResponse(
    output: OutputStream,
    statusCode: Int,
    statusMessage: String,
    headers: Map<String, String>,
    body: ByteArray
  ) {
    val sb = StringBuilder()
    sb.append("HTTP/1.1 $statusCode $statusMessage\r\n")
    sb.append("Content-Length: ${body.size}\r\n")
    sb.append("Connection: close\r\n")

    // Include relevant headers
    val headersToInclude = listOf("Content-Type", "Cache-Control", "Accept-Ranges", "Content-Range")
    for (header in headersToInclude) {
      headers[header]?.let { sb.append("$header: $it\r\n") }
    }

    sb.append("Access-Control-Allow-Origin: *\r\n")
    sb.append("\r\n")

    output.write(sb.toString().toByteArray(Charsets.UTF_8))
    output.write(body)
    output.flush()
  }

  private fun sendError(output: OutputStream, statusCode: Int, message: String) {
    val body = message.toByteArray(Charsets.UTF_8)
    val response = """
      HTTP/1.1 $statusCode $message
      Content-Type: text/plain
      Content-Length: ${body.size}
      Connection: close

      $message
    """.trimIndent()

    output.write(response.toByteArray(Charsets.UTF_8))
    output.flush()
  }

  // MARK: - Manifest Rewriting

  private fun rewriteManifest(data: ByteArray, baseUrlString: String): ByteArray {
    val manifest = String(data, Charsets.UTF_8)
    val baseUrl = try {
      URI(baseUrlString)
    } catch (e: Exception) {
      return data
    }

    val proxyBase = "http://127.0.0.1:$port/proxy?url="
    val lines = manifest.split("\n")
    val rewrittenLines = mutableListOf<String>()

    for (line in lines) {
      val trimmed = line.trim()

      when {
        trimmed.isEmpty() -> rewrittenLines.add(line)

        trimmed.startsWith("#") && trimmed.contains("URI=\"") -> {
          rewrittenLines.add(rewriteUriAttributes(line, baseUrl, proxyBase))
        }

        trimmed.startsWith("#") -> rewrittenLines.add(line)

        else -> {
          val proxyUrl = createProxyUrl(trimmed, baseUrl, proxyBase)
          rewrittenLines.add(proxyUrl ?: line)
        }
      }
    }

    Log.d(TAG, "Rewrote HLS manifest URLs (${lines.size} lines)")
    return rewrittenLines.joinToString("\n").toByteArray(Charsets.UTF_8)
  }

  private fun rewriteUriAttributes(line: String, baseUrl: URI, proxyBase: String): String {
    val pattern = Regex("URI=\"([^\"]+)\"")
    return pattern.replace(line) { match ->
      val uri = match.groupValues[1]
      val proxyUrl = createProxyUrl(uri, baseUrl, proxyBase)
      "URI=\"${proxyUrl ?: uri}\""
    }
  }

  private fun createProxyUrl(urlString: String, baseUrl: URI, proxyBase: String): String? {
    val trimmed = urlString.trim()

    val absoluteUrl = try {
      if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) {
        trimmed
      } else {
        baseUrl.resolve(trimmed).toString()
      }
    } catch (e: Exception) {
      return null
    }

    val encoded = try {
      URLEncoder.encode(absoluteUrl, "UTF-8")
    } catch (e: Exception) {
      return null
    }

    return proxyBase + encoded
  }
}

// Extension to convert OkHttp Headers to Map
private fun okhttp3.Headers.toMap(): Map<String, String> {
  val map = mutableMapOf<String, String>()
  for (i in 0 until size) {
    map[name(i)] = value(i)
  }
  return map
}
